presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/TestRun.java

```diff
@@ -129,20 +141,21 @@ public Execution(EnvironmentFactory environmentFactory, PathResolver pathResolve
             this.testArguments = ImmutableList.copyOf(requireNonNull(testRunOptions.testArguments, "testOptions.testArguments is null"));
             this.environment = requireNonNull(testRunOptions.environment, "testRunOptions.environment is null");
             this.reportsDir = testRunOptions.reportsDir;
+            this.startupTimeout = requireNonNull(testRunOptions.startupTimeout, "testRunOptions.startupTimeout is null");
+            this.startupRetries = testRunOptions.startupRetries;
         }
 
         @Override
         public void run()
         {
             try (UncheckedCloseable ignore = this::cleanUp) {
-                log.info("Pruning old environment(s)");
-                Environments.pruneEnvironment();
-
-                Environment environment = getEnvironment();
-
-                log.info("Starting the environment '%s'", environment);
-                environment.start();
-                log.info("Environment '%s' started", environment);
+                Environment environment = Failsafe.with(new RetryPolicy<>()
+                        .withMaxRetries(startupRetries)
+                        .withBackoff(100, 300000, ChronoUnit.MILLIS) // exponential backoff from 100 ms to 5 min
+                        .onFailedAttempt(event -> log.warn("Could not start environment '%s': %s", this.environment, event.getLastFailure()))
+                        .onRetry(event -> log.info("Trying to start environment '%s', %d failed attempt(s)", this.environment, event.getAttemptCount() + 1))
+                        .onSuccess(event -> log.info("Environment '%s' started in %s, %d attempt(s)", this.environment, event.getElapsedTime(), event.getAttemptCount())))
+                        .get(() -> tryStartEnvironment(startupTimeout));
```

Comment:
Don't you need to setup max duration for failsafe?

Reply:
Not exactly. I have several attempts + timeout for a single startup. I'll add max duration to the failsafe as well